File: src/main/java/org/ldk/structs/Verification.java
package org.ldk.structs;

import org.ldk.impl.bindings;
import org.ldk.enums.*;
import org.ldk.util.*;
import java.util.Arrays;
import java.lang.ref.Reference;
import javax.annotation.Nullable;

/**
 * A trait defining behavior for creating and verifing the HMAC for authenticating a given data.
 */
@SuppressWarnings("unchecked") // We correctly assign various generic arrays
public class Verification extends CommonBase {
	final bindings.LDKVerification bindings_instance;
	Verification(Object _dummy, long ptr) { super(ptr); bindings_instance = null; }
	private Verification(bindings.LDKVerification arg) {
		super(bindings.LDKVerification_new(arg));
		this.ptrs_to.add(arg);
		this.bindings_instance = arg;
	}
	@Override @SuppressWarnings("deprecation")
	protected void finalize() throws Throwable {
		if (ptr != 0) { bindings.Verification_free(ptr); } super.finalize();
	}
	/**
	 * Destroys the object, freeing associated resources. After this call, any access
	 * to this object may result in a SEGFAULT or worse.
	 *
	 * You should generally NEVER call this method. You should let the garbage collector
	 * do this for you when it finalizes objects. However, it may be useful for types
	 * which represent locks and should be closed immediately to avoid holding locks
	 * until the GC runs.
	 */
	public void destroy() {
		if (ptr != 0) { bindings.Verification_free(ptr); }
		ptr = 0;
	}
	public static interface VerificationInterface {
		/**
		 * Constructs an HMAC to include in [`OffersContext`] for the data along with the given
		 * [`Nonce`].
		 */
		byte[] hmac_for_offer_payment(Nonce nonce, ExpandedKey expanded_key);
		/**
		 * Authenticates the data using an HMAC and a [`Nonce`] taken from an [`OffersContext`].
		 */
		Result_NoneNoneZ verify_for_offer_payment(byte[] hmac, Nonce nonce, ExpandedKey expanded_key);
	}
	private static class LDKVerificationHolder { Verification held; }
	public static Verification new_impl(VerificationInterface arg) {
		final LDKVerificationHolder impl_holder = new LDKVerificationHolder();
		impl_holder.held = new Verification(new bindings.LDKVerification() {
			@Override public byte[] hmac_for_offer_payment(long nonce, long expanded_key) {
				org.ldk.structs.Nonce nonce_hu_conv = null; if (nonce < 0 || nonce > 4096) { nonce_hu_conv = new org.ldk.structs.Nonce(null, nonce); }
				if (nonce_hu_conv != null) { nonce_hu_conv.ptrs_to.add(this); };
				org.ldk.structs.ExpandedKey expanded_key_hu_conv = null; if (expanded_key < 0 || expanded_key > 4096) { expanded_key_hu_conv = new org.ldk.structs.ExpandedKey(null, expanded_key); }
				byte[] ret = arg.hmac_for_offer_payment(nonce_hu_conv, expanded_key_hu_conv);
				Reference.reachabilityFence(arg);
				byte[] result = InternalUtils.check_arr_len(ret, 32);
				return result;
			}
			@Override public long verify_for_offer_payment(byte[] hmac, long nonce, long expanded_key) {
				org.ldk.structs.Nonce nonce_hu_conv = null; if (nonce < 0 || nonce > 4096) { nonce_hu_conv = new org.ldk.structs.Nonce(null, nonce); }
				if (nonce_hu_conv != null) { nonce_hu_conv.ptrs_to.add(this); };
				org.ldk.structs.ExpandedKey expanded_key_hu_conv = null; if (expanded_key < 0 || expanded_key > 4096) { expanded_key_hu_conv = new org.ldk.structs.ExpandedKey(null, expanded_key); }
				Result_NoneNoneZ ret = arg.verify_for_offer_payment(hmac, nonce_hu_conv, expanded_key_hu_conv);
				Reference.reachabilityFence(arg);
				long result = ret.clone_ptr();
				return result;
			}
		});
		return impl_holder.held;
	}
	/**
	 * Constructs an HMAC to include in [`OffersContext`] for the data along with the given
	 * [`Nonce`].
	 */
	public byte[] hmac_for_offer_payment(org.ldk.structs.Nonce nonce, org.ldk.structs.ExpandedKey expanded_key) {
		byte[] ret = bindings.Verification_hmac_for_offer_payment(this.ptr, nonce.ptr, expanded_key.ptr);
		Reference.reachabilityFence(this);
		Reference.reachabilityFence(nonce);
		Reference.reachabilityFence(expanded_key);
		return ret;
	}

	/**
	 * Authenticates the data using an HMAC and a [`Nonce`] taken from an [`OffersContext`].
	 */
	public Result_NoneNoneZ verify_for_offer_payment(byte[] hmac, org.ldk.structs.Nonce nonce, org.ldk.structs.ExpandedKey expanded_key) {
		long ret = bindings.Verification_verify_for_offer_payment(this.ptr, InternalUtils.check_arr_len(hmac, 32), nonce.ptr, expanded_key.ptr);
		Reference.reachabilityFence(this);
		Reference.reachabilityFence(hmac);
		Reference.reachabilityFence(nonce);
		Reference.reachabilityFence(expanded_key);
		if (ret >= 0 && ret <= 4096) { return null; }
		Result_NoneNoneZ ret_hu_conv = Result_NoneNoneZ.constr_from_ptr(ret);
		return ret_hu_conv;
	}

}

File: c_sharp/src/org/ldk/structs/Verification.cs

using org.ldk.impl;
using org.ldk.enums;
using org.ldk.util;
using System;

namespace org { namespace ldk { namespace structs {



/** An implementation of Verification */
public interface VerificationInterface {
	/**Constructs an HMAC to include in [`OffersContext`] for the data along with the given
	 * [`Nonce`].
	 */
	byte[] hmac_for_offer_payment(org.ldk.structs.Nonce nonce, org.ldk.structs.ExpandedKey expanded_key);
	/**Authenticates the data using an HMAC and a [`Nonce`] taken from an [`OffersContext`].
	 */
	Result_NoneNoneZ verify_for_offer_payment(byte[] hmac, org.ldk.structs.Nonce nonce, org.ldk.structs.ExpandedKey expanded_key);
}

/**
 * A trait defining behavior for creating and verifing the HMAC for authenticating a given data.
 */
public class Verification : CommonBase {
	internal bindings.LDKVerification bindings_instance;
	internal long instance_idx;

	internal Verification(object _dummy, long ptr) : base(ptr) { bindings_instance = null; }
	~Verification() {
		if (ptr != 0) { bindings.Verification_free(ptr); }
	}

	private class LDKVerificationHolder { internal Verification held; }
	private class LDKVerificationImpl : bindings.LDKVerification {
		internal LDKVerificationImpl(VerificationInterface arg, LDKVerificationHolder impl_holder) { this.arg = arg; this.impl_holder = impl_holder; }
		private VerificationInterface arg;
		private LDKVerificationHolder impl_holder;
		public long hmac_for_offer_payment(long _nonce, long _expanded_key) {
			org.ldk.structs.Nonce _nonce_hu_conv = null; if (_nonce < 0 || _nonce > 4096) { _nonce_hu_conv = new org.ldk.structs.Nonce(null, _nonce); }
			if (_nonce_hu_conv != null) { _nonce_hu_conv.ptrs_to.AddLast(this); };
			org.ldk.structs.ExpandedKey _expanded_key_hu_conv = null; if (_expanded_key < 0 || _expanded_key > 4096) { _expanded_key_hu_conv = new org.ldk.structs.ExpandedKey(null, _expanded_key); }
			byte[] ret = arg.hmac_for_offer_payment(_nonce_hu_conv, _expanded_key_hu_conv);
				GC.KeepAlive(arg);
			long result = InternalUtils.encodeUint8Array(InternalUtils.check_arr_len(ret, 32));
			return result;
		}
		public long verify_for_offer_payment(long _hmac, long _nonce, long _expanded_key) {
			byte[] _hmac_conv = InternalUtils.decodeUint8Array(_hmac);
			org.ldk.structs.Nonce _nonce_hu_conv = null; if (_nonce < 0 || _nonce > 4096) { _nonce_hu_conv = new org.ldk.structs.Nonce(null, _nonce); }
			if (_nonce_hu_conv != null) { _nonce_hu_conv.ptrs_to.AddLast(this); };
			org.ldk.structs.ExpandedKey _expanded_key_hu_conv = null; if (_expanded_key < 0 || _expanded_key > 4096) { _expanded_key_hu_conv = new org.ldk.structs.ExpandedKey(null, _expanded_key); }
			Result_NoneNoneZ ret = arg.verify_for_offer_payment(_hmac_conv, _nonce_hu_conv, _expanded_key_hu_conv);
				GC.KeepAlive(arg);
			long result = ret.clone_ptr();
			return result;
		}
	}

	/** Creates a new instance of Verification from a given implementation */
	public static Verification new_impl(VerificationInterface arg) {
		LDKVerificationHolder impl_holder = new LDKVerificationHolder();
		LDKVerificationImpl impl = new LDKVerificationImpl(arg, impl_holder);
		long[] ptr_idx = bindings.LDKVerification_new(impl);

		impl_holder.held = new Verification(null, ptr_idx[0]);
		impl_holder.held.instance_idx = ptr_idx[1];
		impl_holder.held.bindings_instance = impl;
		return impl_holder.held;
	}

	/**
	 * Constructs an HMAC to include in [`OffersContext`] for the data along with the given
	 * [`Nonce`].
	 */
	public byte[] hmac_for_offer_payment(org.ldk.structs.Nonce nonce, org.ldk.structs.ExpandedKey expanded_key) {
		long ret = bindings.Verification_hmac_for_offer_payment(this.ptr, nonce.ptr, expanded_key.ptr);
		GC.KeepAlive(this);
		GC.KeepAlive(nonce);
		GC.KeepAlive(expanded_key);
		if (ret >= 0 && ret <= 4096) { return null; }
		byte[] ret_conv = InternalUtils.decodeUint8Array(ret);
		return ret_conv;
	}

	/**
	 * Authenticates the data using an HMAC and a [`Nonce`] taken from an [`OffersContext`].
	 */
	public org.ldk.structs.Result_NoneNoneZ verify_for_offer_payment(byte[] hmac, org.ldk.structs.Nonce nonce, org.ldk.structs.ExpandedKey expanded_key) {
		long ret = bindings.Verification_verify_for_offer_payment(this.ptr, InternalUtils.encodeUint8Array(InternalUtils.check_arr_len(hmac, 32)), nonce.ptr, expanded_key.ptr);
		GC.KeepAlive(this);
		GC.KeepAlive(hmac);
		GC.KeepAlive(nonce);
		GC.KeepAlive(expanded_key);
		if (ret >= 0 && ret <= 4096) { return null; }
		Result_NoneNoneZ ret_hu_conv = Result_NoneNoneZ.constr_from_ptr(ret);
		return ret_hu_conv;
	}

}
} } }
